pubtools/_quay/push_docker.py

```diff
@@ -415,26 +416,77 @@ def run(self):
             container_pusher.push_container_images()
 
             if operator_push_items:
                 # Build index images
                 operator_pusher = OperatorPusher(operator_push_items, self.target_settings)
+                existing_index_images = operator_pusher.get_existing_index_images(
+                    self.quay_api_client
+                )
                 iib_results = operator_pusher.build_index_images()
                 # Sign operator images
                 operator_signature_handler = OperatorSignatureHandler(
                     self.hub, self.task_id, self.target_settings
                 )
-                operator_signature_handler.sign_operator_images(iib_results)
+                operator_manifest_claims, _ = operator_signature_handler.sign_operator_images(
+                    iib_results
+                )
                 # Push index images to Quay
                 operator_pusher.push_index_images(iib_results)
         except Exception as e:
             LOG.error("An exception has occurred during the push, starting rollback")
             self.rollback(backup_tags, rollback_tags)
             raise e
+        else:
+            # Remove old signatures
+
+            new_signatures = [
+                (m["manifest_digest"], m["docker_reference"]) for m in manifest_claims
+            ]
+            outdated_signatures = []
+
+            for image_data, manifest in backup_tags.items():
+                ext_repo = image_data.repo.split("/")[1].replace(INTERNAL_DELIMITER, "/")
+                if "manifests" in manifest:
+                    for arch_manifest in manifest["manifests"]:
+                        outdated_signatures.append((arch_manifest, image_data.tag, ext_repo))
+                else:
+                    outdated_signatures.append((manifest["digest"], image_data.tag, ext_repo))
+
+            signatures_to_remove = []
+            for esig in container_signature_handler.get_signatures_from_pyxis(
+                [sig[0] for sig in outdated_signatures]
+            ):
+                if (
+                    esig["manifest_digest"],
+                    esig["reference"].split(":")[-1],
+                    esig["repository"],
+                ) in outdated_signatures and (
+                    esig["manifest_digest"],
+                    esig["reference"],
+                ) not in new_signatures:
+                    signatures_to_remove.append((esig["manifest_digest"], esig["reference"]))
+
+            container_signature_handler.remove_outdated_signatures(signatures_to_remove)
+
+            signatures_to_remove = []
+            if operator_push_items:
+                outdated_signatures = []
+                new_operator_signatures = [
+                    (m["manifest_digest"], m["docker_reference"]) for m in manifest_claims
+                ]
+                for esig in container_signature_handler.get_signatures_from_pyxis(
```

Comment:
Unless we decide to use signature IDs I feel like there's no need to actually gather existing signatures

Reply:
Main reasons for that are: 1. avoid removal of non existing signatures, remove also signature for reference registries which doesn't have to be in target settings (because they were removed for example).